controller/node_controller.go

```diff
@@ -1027,60 +1031,66 @@ func (nc *NodeController) syncMultipathd(node *longhorn.Node, namespaces []lhtyp
+		return
+	}
+
+	if len(notFoundModulesUsingkmod) == 0 {
+		node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeModulesLoaded, longhorn.ConditionStatusTrue, "", "")
+		return
+	}
+
+	notLoadedModules, err := checkModulesLoadedByConfigFile(nc.logger, notFoundModulesUsingkmod, kubeNode.Status.NodeInfo.KernelVersion)
+	if err != nil {
+		node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeModulesLoaded, longhorn.ConditionStatusFalse,
+			string(longhorn.NodeConditionReasonCheckKernelConfigFailed),
+			fmt.Sprintf("Failed to check kernel config file for kernel modules %v: %v", notFoundModulesUsingkmod, err.Error()))
+		return
+	}
+
+	if len(notLoadedModules) != 0 {
+		node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeModulesLoaded, longhorn.ConditionStatusFalse,
+			string(longhorn.NodeConditionReasonModulesNotLoaded),
+			fmt.Sprintf("Modules %v are not loaded on node %v", notLoadedModules, node.Name))
 		return
 	}
 
-	kernelConfigPath := "/boot/config-" + kernelVersion
-	args := []string{kernelConfigPath}
-	if _, err := nsexec.Execute(nil, "ls", args, lhtypes.ExecuteDefaultTimeout); err != nil {
+	node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeModulesLoaded, longhorn.ConditionStatusTrue, "", "")
+}
+
+func (nc *NodeController) syncNFSClientVersion(kubeNode *corev1.Node, node *longhorn.Node, namespaces []lhtypes.Namespace) {
+	kernelVersion := kubeNode.Status.NodeInfo.KernelVersion
+	nfsClientVersions := []string{"CONFIG_NFS_V4_2", "CONFIG_NFS_V4_1", "CONFIG_NFS_V4"}
+
+	kernelConfigPath := kernelConfigFilePathPrefix + kernelVersion
+	configContent, err := lhio.ReadFileContent(kernelConfigPath)
+	if err != nil {
 		node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeNFSClientInstalled, longhorn.ConditionStatusFalse,
 			string(longhorn.NodeConditionReasonKernelConfigIsNotFound),
 			fmt.Sprintf("Unable to find %v for checking %v: %v", kernelConfigPath, nfsClientVersions, err.Error()))
 		return
 	}
 
 	for _, ver := range nfsClientVersions {
-		args := []string{ver + "=", kernelConfigPath}
-		result, err := nsexec.Execute(nil, "grep", args, lhtypes.ExecuteDefaultTimeout)
+		moduleEnabled, err := checkKernelModuleEabled(nc.logger, configContent, ver, "nfs")
 		if err != nil {
-			nc.logger.WithError(err).Debugf("Failed to find kernel config %v on node %v", ver, node.Name)
-			continue
+			node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeModulesLoaded, longhorn.ConditionStatusFalse,
+				string(longhorn.NodeConditionReasonNamespaceExecutorErr),
+				fmt.Sprintf("Failed to check kernel module %v: %v", ver, err.Error()))
+			return
 		}
-		enabled := strings.TrimSpace(strings.Split(result, "=")[1])
-		switch enabled {
-		case "y":
+		if moduleEnabled {
```

Comment:
Can we directly use `checkModulesLoadedByConfigFile` instead? The input `modules` would be `map[string]string{"nfs": "CONFIG_NFS_V4_2", "nfs": "CONFIG_NFS_V4_1", "nfs": "CONFIG_NFS_V4"}`

Reply:
Yes, it is modified. Thanks.